### PR TITLE
Matching Engine V2 new PR on top of the part 1 PR#1225 with already approved code

### LINKED ISSUE #1222: Hyperdrive Matching Engine
# Hyperdrive Matching Engine

## Background

In order to support large fixed borrow transactions, it is important that Hyperdrive can support the direct matching of long and short traders. #1201 added an extension contract that makes direct matching possible by using flash loans to add liquidity right before the trades are executed, but this approach has several issues. First, the orders must line up perfectly to ensure that the flash loan can be repaid, and these parameters are quite brittle. If even a small trade is executed before the orders are matched, the add liquidity amount will need to be adjusted, which is currently a manual and time consuming process. The second issue with the system is that it relies on flash loan liquidity. This places an upper-bound on the amount of volume that can feasibly flow through the system. The third issue is that Hyperdrive trades aren't linear, so the current matching engine system isn't set up to support partial fills. 

We can solve all of these issues by making it possible for the matching engine to circumvent the AMM logic. Engineering is underway for `mint` and `burn` functions in #1220 that will make it easier to develop new execution mechanisms for Hyperdrive. `mint` takes in base or vault shares that will underly a set of bonds. Once the `bondAmount` has been calculated, `mint` will create an equal amount of long and short tokens in the current checkpoint and send the proceeds to a `longDestination` and `shortDestination`. `burn` does the opposite. Given an equal amount of long and short tokens in the same checkpoint, `burn` will release the capital underlying the bonds. Using these functions, we can build a matching engine that is simpler and also more powerful. 

Our new matching engine should plug into `mint` and `burn` to support trades of multiple different types. Traders should be able to post their intent to buy or sell either the long or the short position in different checkpoints. Different combinations of these orders can be matched to give a pair of traders what they each want.

## Requirements

The goal of this task is to build a matching engine with the following requirements:

1. A simple schema for order intents that specifies:
    - The size of the trade in bonds
    - The price or fixed rate the trader should receive
    - The position they are buying or selling (long or short)
    - The asset they are using to buy or sell (base or vault shares)
    - The order's expiry timestamp.
    - The Hyperdrive pool
    - The direction of the trade (buy or sell)
2. Order validation logic that verifies:
    - The orders signature. This should support EOA signatures and EIP1271 signatures.
    - If the order has expired or not.
    - If the order has been cancelled or filled.
3. A function to cancel orders
4. A function to match orders
    - This should support all the different types of matches (buy long <> sell long, buy short <> sell short, buy long <> buy short, and sell long <> sell short)
    - Supports partial matches
    - Supports ETH and ERC20 tokens
    - Think about how to support different combintations of `asBase = true` and `asBase = false`.

# Examples
     
## Ex. Imperfect Match (price matches, size doesn't)

Order 1: Alex wants to buy 1000 longs for 990 USDC
Order 2: Sean wants to buy 500 shorts for 5 USDC
     
We will partially fill Alex's order and fully fill Sean's order.

Alex will get 500 longs for 495 USDC
Sean will get 500 shorts for 5 USDC

495 + 5 = 500 -> mint will mint 500 longs and 500 shorts

### Ex 1. Feasible

Order 1: Alex wants to buy 1000 longs for 990 USDC
Order 2: Sean wants to buy 1000 shorts for 11 USDC

990 + 11 > 1000, they can both be matched

Alex gets 1000 longs for 990 USDC
Sean gets 1000 shorts for 11 USDC

There is 1 leftover USDC. This can go to either Alex, Sean, or the person that calls `match`

### Ex 2. Infeasible

Order 1: Alex wants to buy 1000 longs for 990 USDC
Order 2: Sean wants to buy 1000 shorts for 9 USDC

This doesn't work. 990 + 9 = 999 < 1000, so we don't have enough money to mint the bonds

990 * (999 / 1000) = 989.01
9 * (999 / 1000) = 8.991

989.01 + 8.991 = 998.001 < 999

## Different types of orders

The following situations shows the different kinds of orders that can be matched. Some types of matches require each trader to have USDC whereas other types of matches require one or both traders have long or short positions.
 
Order 1: Alex wants 1000 longs for 990 USDC
Order 2: Sean wants 1000 shorts for 10 USDC

Order 1: Alex wants 1000 longs for 990 USDC
Order 2: Sean wants 990 USDC for 1000 longs 

Order 1: Alex wants 1000 shorts for 10 USDC
Order 2: Sean wants 10 USDC for 1000 shorts 

Order 1: Alex wants 990 USDC for 1000 longs
Order 2: Sean wants 10 USDC for 1000 shorts

### PR DESCRIPTION
**Pull request by Sean329**
# HyperdriveMatchingEngineV2 PR Description

## Overview
This PR introduces `HyperdriveMatchingEngineV2`, an enhanced matching engine for Hyperdrive that facilitates efficient order matching and position trading. The engine supports direct position minting/burning and position transfers between traders.

### Order Types Support
- **OpenLong**: Open a long position
- **OpenShort**: Open a short position
- **CloseLong**: Close an existing long position
- **CloseShort**: Close an existing short position

### Core Functionalities
1. **Order Matching (`matchOrders`)**
   - Matches complementary orders (e.g., OpenLong + OpenShort)
   - Supports position transfers between traders
   - Handles partial fills
   - Includes surplus distribution mechanism

2. **Direct Order Filling (`fillOrder`)**
   - Allows immediate order execution by takers
   - Supports maker-taker model
   - Maintains order state tracking

3. **EIP-712 Signature Support**
   - Secure order signing and verification
   - Support for both EOA and contract signatures

### Safety Features
- Reentrancy protection
- Order expiration checks
- Counterparty validation
- Maturity time validation
- Fund amount validation
- Token amount buffer for gas optimization

## Technical Improvements
- Direct mint/burn operations instead of flash loans
- Optimized gas usage with TOKEN_AMOUNT_BUFFER
- Enhanced order state tracking
- Comprehensive error handling

## Testing Coverage
The implementation includes extensive test coverage:
- Basic order matching scenarios
- Position transfer cases
- Partial fill scenarios
- Edge cases and failure conditions
- Fuzzing tests for TOKEN_AMOUNT_BUFFER validation

## Interface Changes
The `IHyperdriveMatchingEngineV2` interface provides:
- Clear error definitions
- Comprehensive event logging
- Structured order intent definition
- Order amount tracking

## Security Considerations
- All monetary operations use SafeERC20
- Strict validation of order parameters
- Protected against common attack vectors
- Clear access control mechanisms

## Breaking Changes
None. This is a new contract implementation that can coexist with previous versions.

## Next Steps
- [ ] Complete security audit
- [ ] Add integration tests
- [ ] Update documentation
- [ ] Deploy to testnet

## Related Issues
- Closes #[issue-number]
- References #[related-issue-number]

resolves #1222

### CODE REVIEW

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol [hunk 1]
```diff
@@ -72,154 +69,151 @@ contract HyperdriveMatchingEngineV2 is
     /// @notice Matches two orders.
```
Comment (jalextowle) [inline contracts/src/matching/HyperdriveMatchingEngineV2.sol:69]: I think we need some more documentation about how the orders are structured in the different matching situations. The idea that `order1` needs to be `OpenLong` if `order2` is `OpenShort` is pretty opinionated. I honestly think it would be easy enough to handle both cases and just record the `longOrder` and `shortOrder` at the start of the if-statement, but if we're going to stick with this kind of implementation, then it needs to be spelled out clearly for integrators.

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol [hunk 2]
```diff
@@ -228,56 +222,607 @@ contract HyperdriveMatchingEngineV2 is
...
 
     /// @notice Allows traders to cancel their orders.
     /// @param _orders Array of orders to cancel.
-    function cancelOrders(OrderIntent[] calldata _orders) external nonReentrant {
+    function cancelOrders(
+        OrderIntent[] calldata _orders
+    ) external nonReentrant {
         bytes32[] memory orderHashes = new bytes32[](_orders.length);
-        
+
```
Comment (jalextowle) [inline contracts/src/matching/HyperdriveMatchingEngineV2.sol:825]: Nit: Either remove this or add a comment per the style guide

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol [hunk 3]
```diff
@@ -605,41 +1245,183 @@ contract HyperdriveMatchingEngineV2 is
...
+        //      first, then transfer the needed amount to the close trader.
+        _fundToken.safeTransferFrom(
+            _openOrder.trader,
+            address(this),
+            _fundTokenAmountOpenOrder
+        );
+        _fundToken.safeTransfer(
+            _closeOrder.options.destination,
+            _minFundAmountCloseOrder
+        );
```
Comment (jalextowle) [inline contracts/src/matching/HyperdriveMatchingEngineV2.sol:1335]: It would be quite a bit more efficient if we just did the transfer directly. I get what you're trying to do, but the likelihood that there are donation tokens feels low enough to not warrant the additional cost IMO

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol [hunk 4]
```diff
@@ -605,41 +1245,183 @@ contract HyperdriveMatchingEngineV2 is
...
+        IHyperdrive.PoolConfig memory config = _hyperdrive.getPoolConfig();
+
+        // Calculate checkpoint and maturity time.
+        uint256 latestCheckpoint = _latestCheckpoint(config.checkpointDuration);
+        maturityTime = latestCheckpoint + config.positionDuration;
+
+        // Get vault share prices.
+        // @dev TODO: there is another way to get the info without calling
+        //      getPoolInfo()?
+        uint256 vaultSharePrice = _hyperdrive.getPoolInfo().vaultSharePrice;
```
Comment (jalextowle) [inline contracts/src/matching/HyperdriveMatchingEngineV2.sol:1369]: Yes, you can call `hyperdrive.convertToBase(ONE)`.

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol (hunk 1 above)
Comment (Sean329) [reply contracts/src/matching/HyperdriveMatchingEngineV2.sol:69]: Spelled out documentation.

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol (hunk 3 above)
Comment (Sean329) [reply contracts/src/matching/HyperdriveMatchingEngineV2.sol:1335]: Agreed

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol (hunk 4 above)
Comment (Sean329) [reply contracts/src/matching/HyperdriveMatchingEngineV2.sol:1369]: Is ONE always supposed to be 1e18? What if the base is 1e6 like USDC?

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol (hunk 3 above)
Comment (Sean329) [reply contracts/src/matching/HyperdriveMatchingEngineV2.sol:1335]: Actually, I just recalled why I did it this way --- even if we directly transfer the `_minFundAmountCloseOrder` from the openOrder trader to the closeOrder destination, we still need to transfer any surplus amount `_fundTokenAmountOpenOrder  - _minFundAmountCloseOrder` into `address(this)` because the handling of surplus to the caller or taker happens in the upper level function call. Therefore, we need 2 transfers no matter how.

File: contracts/src/matching/HyperdriveMatchingEngineV2.sol (hunk 4 above)
Comment (Sean329) [reply contracts/src/matching/HyperdriveMatchingEngineV2.sol]: Implemented

**Review by jalextowle — APPROVED** [review]
